cmd/encryption.go

```diff
@@ -0,0 +1,186 @@
+
+	"github.com/minio/sio"
+)
+
+// AWS errors for invalid SSE-C requests.
+var (
+	errInsecureSSERequest  = errors.New("Requests specifying Server Side Encryption with Customer provided keys must be made over a secure connection")
+	errInvalidSSEAlgorithm = errors.New("Requests specifying Server Side Encryption with Customer provided keys must provide a valid encryption algorithm")
+	errMissingSSEKey       = errors.New("Requests specifying Server Side Encryption with Customer provided keys must provide an appropriate secret key")
+	errInvalidSSEKey       = errors.New("The secret key was invalid for the specified algorithm")
+	errMissingSSEKeyMD5    = errors.New("Requests specifying Server Side Encryption with Customer provided keys must provide the client calculated MD5 of the secret key")
+	errSSEKeyMD5Mismatch   = errors.New("The calculated MD5 hash of the key did not match the hash that was provided")
+)
+
+const (
+	SSECustomerAlgorithm = "X-Amz-Server-Side-Encryption-Customer-Algorithm"
+	SSECustomerKey       = "X-Amz-Server-Side-Encryption-Customer-Key"
+	SSECustomerKeyMD5    = "X-Amz-Server-Side-Encryption-Customer-Key-MD5"
+)
+
+// SSECustomerKeySize is the size of valid client provided encryption keys in bytes.
+// Currently AWS supports only AES256. So the SSE-C key size is fixed to 32 bytes.
+const SSECustomerKeySize = 32
+
+// IsSSECustomerRequest returns true if the given HTTP header
+// contains server-side-encryption with customer provided key fields.
+func IsSSECustomerRequest(header http.Header) bool {
+	if header.Get(SSECustomerAlgorithm) != "" {
+		return true
+	}
+	if header.Get(SSECustomerKey) != "" {
+		return true
+	}
+	if header.Get(SSECustomerKeyMD5) != "" {
+		return true
+	}
+	return false
+}
+
+// ParseSSECustomerRequest parses the SSE-C header fields of the provided header.
+// It returns the client provided key on success.
+func ParseSSECustomerRequest(header http.Header) (key []byte, err error) {
+	if algorithm := header.Get(SSECustomerAlgorithm); algorithm != "AES256" { // Currently AWS supports only AES256
+		return nil, errInvalidSSEAlgorithm
+	}
+	if header.Get(SSECustomerKey) == "" {
+		return nil, errMissingSSEKey
+	}
+	if header.Get(SSECustomerKeyMD5) == "" {
+		return nil, errMissingSSEKeyMD5
+	}
+
+	key, err = base64.StdEncoding.DecodeString(header.Get(SSECustomerKey))
+	if err != nil {
+		return nil, errInvalidSSEKey
+	}
+	if len(key) != SSECustomerKeySize {
+		return nil, errInvalidSSEKey
+	}
+
+	keyMD5, err := base64.StdEncoding.DecodeString(header.Get(SSECustomerKeyMD5))
+	if err != nil {
+		return nil, errSSEKeyMD5Mismatch
+	}
+	if md5Sum := md5.Sum(key); !bytes.Equal(md5Sum[:], keyMD5) {
+		return nil, errSSEKeyMD5Mismatch
+	}
+	return key, nil
+}
+
+// RangeWriter returns an io.Writer which ignores all data written
+// to it before start and all data written to it after end.
+func RangeWriter(w io.Writer, start, end int64) io.Writer {
+	if start < 0 {
+		panic("start cannot be negative")
```

Comment:
We shouldn't panic in our server code only under dire situations, especially for argument validation we should simply return errInvalidArgument in this case.